staging/src/k8s.io/csi-api/pkg/apis/csi/v1alpha1/types.go

```diff
@@ -97,8 +107,33 @@ type CSINodeInfo struct {
 	CSIDrivers []CSIDriverInfo `json:"csiDrivers" patchStrategy:"merge" patchMergeKey:"driver"`
 }
 
-// CSIDriverInfo contains information about one CSI driver installed on a node.
+// CSIDriverInfo holds information about one CSI driver installed on a node.
 type CSIDriverInfo struct {
+	// spec is the specification of CSIDriverInfo
+	Spec CSIDriverInfoSpec `json:"spec"`
+
+	// status is the current status of CSIDriverInfo
+	Status CSIDriverInfoStatus `json:"status"`
```

Comment:
It's fairly uncommon to have spec and status embedded in API objects beyond the top level. Is spec used by higher level actors to inform the system of the desired state? If so, would any changes made by the system in reaction to a spec change be reflected in status?

Reply:
Wait, this isn't the top level? Oops, I should have noticed that. Hazards of not expanding the git diff.

Yes, spec and status belong in the root of the object and should have independent lists of drivers.

Reply:
Just to get some early feedback before I make sweeping changes, does this seem about right to you:
```
// CSINodeInfo holds information about all CSI drivers installed on a node.
type CSINodeInfo struct {
	metav1.TypeMeta `json:",inline"`

	// metadata.name must be the Kubernetes node name.
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec is the specification of CSINodeInfo
	// +patchMergeKey=spec
	// +patchStrategy=merge
	Spec CSINodeInfoSpec `json:"spec" patchStrategy:"merge" patchMergeKey:"spec"`

	// status is the current status of CSINodeInfo
	// +patchMergeKey=status
	// +patchStrategy=merge
	Status CSINodeInfoStatus `json:"status" patchStrategy:"merge" patchMergeKey:"status"`
}

type CSINodeInfoSpec struct {
	// csiDriverInfoSpecs is a list of specifications of CSIDriverInfo
	CSIDriverInfoSpecs []CSIDriverInfoSpec `json:"csiDriverInfoSpecs"`
}

type CSINodeInfoStatus struct {
	// csiDriverInfoStatuses is a list of the statuses of CSIDriverInfo
	CSIDriverInfoStatuses []CSIDriverInfoStatus `json:"csiDriverInfoStatuses"`
}
```

Reply:
Yes, although I'd probably just say "drivers" in spec & status instead of CSIDriverInfoSpecs / CSIDriverInfoStatuses.

Reply:
I agree with Daniel. However, I want to mention that objects don't need a spec and a status. If the spec isn't used by a higher level actor to direct a reconciling system to modify current state, and status isn't used to echo back that state, it might not make sense. I'm not actually sure how these fields are intended to be used. Does kubelet set both?

Reply:
I agree with Daniel. However, I want to mention that objects don't need a spec and a status. If the spec isn't used by a higher level actor to direct a reconciling system to modify current state, and status isn't used to echo back that state, it might not make sense. I'm not actually sure how these fields are intended to be used. Does kubelet set both?

Reply:
@mikedanese Kubelet currently sets both, and only during plugin registration/de-registration. But there are future plans/requests to have a reconciling system modify the state (particularly "Available") a la https://github.com/kubernetes/kubernetes/issues/67972. As long as having a Spec and a Status isn't "wrong" when we don't have this reconciling system, I'd rather keep it split so as to make potential future usage easier.